blocks/search.py

```diff
@@ -0,0 +1,261 @@
+from theano import function
+from theano import tensor
+
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+class Search(object):
+    """Abstract search class
+
+    Parameters
+    ----------
+    sequence_generator : sequence generator
+        Sequence generator to be used
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, sequence_generator):
+        self.generator = sequence_generator
+        self.compiled = False
+
+    @abstractmethod
+    def compile(self):
+        self.compiled = True
+
+    @abstractmethod
+    def search(self, beam_size, **kwargs):
+        r"""Performs search
+
+        Parameters
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+        """
+        if not self.compiled:
+            self.compile()
+
+
+class GreedySearch(Search):
+
+    def __init__(self, sequence_generator):
+        super(GreedySearch, self).__init__(sequence_generator)
+
+
+class BeamSearch(Search):
+    """Beam search
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam
+    batch_size : int
+        Size of input batch
+    sequence_generator : sequence generator
+        Sequence generator brick
+
+    """
+    def __init__(self, beam_size, batch_size, sequence_generator, attended,
+                 y, attended_mask, y_mask, inputs_dict, inputs):
+        super(BeamSearch, self).__init__(sequence_generator)
+        self.beam_size = beam_size
+        self.batch_size = batch_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.y = y
+        self.attended_mask = attended_mask
+        self.y_mask = y_mask
+        self.inputs_dict = inputs_dict
+        self.inputs = inputs
+        self.generate_names = sequence_generator.generate_outputs()
+        self.init_computer = None
+        self.next_computer = None
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beamsearch
+
+        Parameters
+        ----------
+        inputs : dict
+            Dictionary of named inputs
+
+        """
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
+        init_generated = generator.generate(attended=attended,
+                                            attended_mask=attended_mask,
+                                            iterate=False, n_steps=1,
+                                            batch_size=self.batch_size)
+        init_generated = OrderedDict(zip(self.generate_names, init_generated))
+        init_cg = ComputationGraph(init_generated.values())
+        init_readouts = VariableFilter(application=generator.readout.emit,
+                                       name='readouts')(init_cg.variables)[-1]
+        init_probs = generator.readout.emitter._probs(init_readouts)
```

Comment:
If `probs` is already an application method, why don't you retrieve the output of this application method directly from the graph? Why do you need to fetch readouts first?